src/server/dflycmd.cc

```diff
@@ -119,6 +120,7 @@ void DflyCmd::ReplicaInfo::Cancel() {
     }
 
     flow->full_sync_fb.JoinIfNeeded();
+    flow->conn = nullptr;
```

Comment:
I am pretty sure we have a data race bug in our dflycmd code. 
The bug is that we call "OnClose" for every replicating connection, assuming that it removes ReplicaInfo for the session from the map. 

In reality, another thread could still hold a `replica_ptr` pointer pointing to ReplicaInfo that has been removed. That struct would have flows arrays with `conn` pointer pointing to potentially closed and freed connection. 
This change at least formalizes the state, setting the dangling pointer to null.

Reply:
This why we have the mutex and state checks. Once you atomically modified the state to cancelled, no other operation should be accessing those